contracts/cw4973/src/lib.rs

```diff
@@ -0,0 +1,319 @@
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    token_id: String,
+) -> Result<Response, ContractError> {
+    // create execute message to burn the nft using the nft_id and ExecuteMsg::Burn of cw721-base
+    let burn_msg = Cw721BaseExecuteMsg::Burn { token_id };
+    // burn the nft with the given id
+    match Cw4973Contract::default().execute(deps, _env, info, burn_msg) {
+        Ok(r) => Ok(r),
+        Err(e) => Err(ContractError::Cw721ContractError(e)),
+    }
+}
+
+// _get_bech32_address function is used to get the bech32 address from the public key and hrp
+fn _get_bech32_address(hrp: &str, pubkey: &[u8]) -> Result<String, ContractError> {
+    let pk_hash = Sha256::digest(pubkey);
+    let rip_result = Ripemd160::digest(pk_hash);
+
+    let address_bytes = rip_result.to_base32();
+    let bech32_address = bech32::encode(hrp, address_bytes, Bech32)
+        .map_err(|err| ContractError::Hrp(err.to_string()))?;
+
+    Ok(bech32_address)
+}
+
+// _safeCheckAgreement function is used to check if the agreement is valid or not
+// if the agreement is valid then it returns the id of the nft
+fn _safe_check_agreement(
+    deps: &DepsMut,
+    env: &Env,
+    active: &str,
+    passive: &str,
+    uri: &str,
+    signature: &PermitSignature,
+) -> Result<String, ContractError> {
+    // get chain id from blockinfo of env
+    let chain_id = env.block.chain_id.clone();
+
+    // get hash for the agreement
+    let hash = _get_hash(active, passive, uri, &chain_id);
+
+    // get the signature value from the signature
+    let sig = base64::decode(signature.signature.clone()).unwrap();
+
+    // get the public key from the signature
+    let pubkey = signature.pub_key.clone();
+
+    // decode the public key
+    let pubkey_bytes = base64::decode(pubkey).unwrap();
+
+    // verify the signature using the hash and the public key
+    let is_verified = deps.api.secp256k1_verify(&hash, &sig, &pubkey_bytes);
+
+    match is_verified {
+        Ok(true) => {
+            // If the signature is verified then we must check the address of public key is equal to passive address
+            // get hrp from the signature
+            let hrp = signature.hrp.clone();
+
+            // get the address of signer from the public key using get_bech32_address function
+            let signer_address = _get_bech32_address(&hrp, pubkey_bytes.as_slice()).unwrap();
+
+            if signer_address != *passive {
+                Err(ContractError::InvalidSigner)
+            } else {
+                // return hex encoded hash
+                Ok(hex::encode(hash))
+            }
+        }
+        Ok(false) => Err(ContractError::InvalidSignature),
+        Err(_) => Err(ContractError::CannotVerifySignature),
+    }
+}
+
+// the get_hash funtion will concat the address of the sender, the address of the 'to', the uri of the nft and the hash of the string
+fn _get_hash(active: &str, passive: &str, uri: &str, chain_id: &str) -> Vec<u8> {
```

Comment:
could you expand a little on what these function arguments are? are active, passive, and uri intended to be additional metadata? i'm curious why you include chain id?